### PR TITLE
Torch NN lab

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77425

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/functional/linear.h [hunk 1]
```diff
@@ -6,6 +6,11 @@ namespace torch {
 namespace nn {
 namespace functional {
 
+inline Tensor bias(const Tensor& input, const Tensor& bias){
+  // Do I need to check the constraints here?
+  return torch::bias(input, bias);
```
Comment (drisspg) [inline torch/csrc/api/include/torch/nn/functional/linear.h:11]: by using the torch:: namespace am I calling the native function defined in linear.cpp

For both python and c++ api the calling structure should be? Module -> functional -> native ?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1931,6 +1931,18 @@ def hardsigmoid(input: Tensor, inplace: bool = False) -> Tensor:
         return torch._C._nn.hardsigmoid_(input)
     return torch._C._nn.hardsigmoid(input)
 
+bias = _add_docstr(
+    torch._C._nn.bias,
```
Comment (drisspg) [inline torch/nn/functional.py:1935]: And from reading the native_functions read me this the entrypoint to at::native::bias

File: torch/overrides.py [hunk 3]
```diff
@@ -332,6 +332,7 @@ def get_testing_overrides() -> Dict[Callable, Callable]:
     # See Issue #28233.
     Tensor = torch.Tensor
     ret: Dict[Callable, Callable] = {
+        torch.nn.functional.bias: lambda input, bias: -1,
```
Comment (drisspg) [inline torch/overrides.py:335]: not totally sure if this is the right/only one one needed here

Comment (jbschlosser) [reply torch/overrides.py:335]: I think this is it; it's the only function defined on the Python side

File: torch/testing/_internal/common_nn.py [hunk 4]
```diff
@@ -99,6 +99,15 @@ def get_weight(m):
 
 
 module_tests = [
+    dict(
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:102]: Looks good! FYI (I probably need to update the module lab) this list of dicts is being superseded by a database of ModuleInfo entries for generating module tests (akin to OpInfo if you're familiar with that).

https://github.com/pytorch/pytorch/blob/e011a8e18bf469a6a612fd1e7647159c353730a9/torch/testing/_internal/common_modules.py#L994

**Review by jbschlosser — COMMENTED** [review]
Looks pretty good! Left some comments below purely for informative reasons. We can close this PR as complete once everything is passing :)

File: torch/csrc/api/include/torch/nn/modules/linear.h [hunk 5]
```diff
@@ -13,6 +13,40 @@
...
+ public:
+  BiasImpl(int64_t num_features)
+    : BiasImpl(BiasOptions(num_features)) {}
+  explicit BiasImpl(const BiasOptions& options_);
+
+  void reset() override;
+
+  void reset_parameters();
+
+  /// Pretty prints the `Linear` module into the given `stream`.
```
[the remark this answers is not in the document]
Comment (drisspg) [reply torch/csrc/api/include/torch/nn/modules/linear.h:30]: yup copy pasta

File: torch/csrc/api/src/nn/modules/linear.cpp [hunk 6]
```diff
@@ -13,6 +13,31 @@ namespace F = torch::nn::functional;
...
+  reset();
+}
+
+void BiasImpl::reset() {
+  bias = register_parameter("bias", torch::empty(options.num_features()));
+  reset_parameters();
+}
+
+void BiasImpl::reset_parameters() {
+  torch::nn::init::uniform_(bias); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
```
[the remark this answers is not in the document]
Comment (drisspg) [reply torch/csrc/api/src/nn/modules/linear.cpp:28]: whooops yeah I think this one more than a nit but again copy pasta from linear

**Comment by drisspg** [comment]
Appears I have two errors which feel related to how I did things:

`pytest test/test_public_bindings.py` ->
```bash
# torch.nn.functional.bias:
  - Is public: it is an attribute that does not start with `_` on a module that does not have `__all__` defined
  - Does NOT look public: because its `__module__` attribute (`torch._C._nn`) is not within the torch library or does not start with the submodule where it is defined (`torch.nn.functional`)
  - You can do either of these two things to fix this problem:
    - To make it NOT public: either define a `__all__` for `torch.nn.functional` or add a `_` at the beginning of the name
    - To make it look public: make sure the `__module__` is properly set and points to a submodule of `torch.nn.functional`
```

and `pytest test/test_fx.py -v -k bias` -> 
```bash
________________________________________________________________________________ TestFunctionalTracing.test_nn_functional_bias _________________________________________________________________________________
Traceback (most recent call last):
  File "/Users/drisspg/meta/pytorch/test/test_fx.py", line 3994, in functional_test
    symbolic_trace(fn)
  File "/Users/drisspg/meta/pytorch/torch/fx/_symbolic_trace.py", line 872, in symbolic_trace
    graph = tracer.trace(root, concrete_args)
  File "/Users/drisspg/meta/pytorch/torch/fx/_symbolic_trace.py", line 550, in trace
    assert isinstance(fn, FunctionType)
AssertionError
```

**Comment by soulitzer** [comment]
Both look like things that fail for `linear` as well, which is currently just manually added to a skip lists. See `test/allowlist_for_publicAPI.json` for the `test_public_bindings`'s `__module__` check, and `UNTRACEABLE_FUNCTIONALS` in `test/test_fx.py` for the second error. 

For the first error, maybe we just want to set `__module__` to what it wants to see? cc @anjali411 
For the second, we can probably just ignore and add to the skip list.

**Comment by drisspg** [comment]
> Both look like things that fail for `linear` as well, which is currently just manually added to a skip lists. See `test/allowlist_for_publicAPI.json` for the `test_public_bindings`'s `__module__` check, and `UNTRACEABLE_FUNCTIONALS` in `test/test_fx.py` for the second error.
> 
> For the first error, maybe we just want to set `__module__` to what it wants to see? cc @anjali411 For the second, we can probably just ignore and add to the skip list.

Thanks for the adding to the Untraceable list fixed the fx error.  I am confused though as to where I would define the __module__ method since the functional definition in python is just this:
```
bias = _add_docstr(
    torch._C._nn.bias,
    r"""
bias(input, bias) -> Tensor

Adds a bias vector the last dimension of input tensor

Shape:
    - Input: math:`(*, num\_features)` where `*` means any number of
      additional dimensions, including none
    - Bias: :math:`(num\_features)` or :math:`()`
    - Output: :math:`(*, num\_features)` where `*` means any number of
      additional dimensions, including none, same shape as Input
""")
```

**Comment by soulitzer** [comment]
Oh the `__module__` attribute is just a string on a function/class that indicates where it was originally defined, so you could just do something like:
```
bias = _add_docstr(
    torch._C._nn.bias, r"""
bias(input, bias) -> Tensor
""")
bias.__module__ = "torch.nn.functional"
```

**Comment by drisspg** [comment]
Ohh nice I tried that out and that appears to make the public binding test. Pass as well. I wonder if there should be another func besides _add_doc_string that does both at init in the module.

**Pull request closed without merging**